File: Data Structures & Algorithms/permutations/submission-2.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        def solve(index, temp):
            if index == len(nums):
                res.append(temp.copy())
                return
            for i in range(index, len(nums)):
                temp[index], temp[i] = temp[i], temp[index]
                solve(index+1, temp)
                temp[index], temp[i] = temp[i], temp[index]
        solve(0, nums)
        return res

File: Data Structures & Algorithms/subsets-ii/submission-2.cpp
class Solution {
public:
    set<vector<int>> st;
    void solve(int index, vector<int> temp, vector<int> nums) {
        if (index == nums.size()) {
            st.insert(temp);
            return;
        }
        temp.push_back(nums[index]);
        solve(index+1, temp, nums);
        temp.pop_back();
        solve(index+1, temp, nums);
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        solve(0, {}, nums);
        vector<vector<int>> ans(st.begin(), st.end());
        return ans;
    }
};
